src/main/java/org/spongepowered/api/map/MapRenderer.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.map;
+
+
+import org.spongepowered.api.entity.living.player.Player;
+
+import javax.annotation.Nullable;
+
+/**
+ * Plugin classes should implement this to modify the image shown on a map.
+ */
+public interface MapRenderer {
+
+    /**
+     * Returns if this map renders different views for each player viewing it.
+     *
+     * <p>Most maps representing shared content are the same for all players
+     * that hold them, but sometimes a plugin may wish to render a unique view
+     * for each player.</p>
+     *
+     * @return True if the rendering is specific to each player, false otherwise
+     */
+    boolean isContextual();
+
+    /**
+     * Gives access to the {@link MapCanvas} for rendering onto a map, this also
+     * provides a viewer and a reference to the original map data to allow
+     * updating cursors and accessing relevant information to the render.
+     *
+     * <p>The viewer is guaranteed to be present if {@link #isContextual()}
+     * returns true.</p>
+     *
+     * @param viewer The player viewing this render, null otherwise
+     * @param view The map that's being rendered
+     * @param canvas The canvas that the renderer will draw onto
+     */
+    void render(@Nullable Player viewer, MapView view, MapCanvas canvas);
```

Comment:
If you followed the API you would see this is the only way to get a
MapCanvas to set those pixels.

On Sat, Apr 2, 2016, 3:38 PM Aaron Hill <notifications@github.com> wrote:

> In src/main/java/org/spongepowered/api/map/MapRenderer.java
> <https://github.com/SpongePowered/SpongeAPI/pull/1162#discussion_r58299549>
> :
>
> > +     */
> > +    boolean isContextual();
> > +
> > +    /**
> > +     * Gives access to the {@link MapCanvas} for rendering onto a map, this also
> > +     * provides a viewer and a reference to the original map data to allow
> > +     * updating cursors and accessing relevant information to the render.
> > +     *
> > +     * <p>The viewer is guaranteed to be present if {@link #isContextual()}
> > +     * returns true.</p>
> > +     *
> > +     * @param viewer The player viewing this render, null otherwise
> > +     * @param view The map that's being rendered
> > +     * @param canvas The canvas that the renderer will draw onto
> > +     */
> > +    void render(@Nullable Player viewer, MapView view, MapCanvas canvas);
>
> When is this called? Under what circimstances would a plugin add a
> MapRenderer instead of directly setting pixels on the map?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/1162/files/d8338269be828e278d15ed5ffbe9bb30b7c899c1#r58299549>
>